src/cudnn/conv.jl

```diff
@@ -10,9 +10,40 @@ using CUDA.CUDNN: scalingParameter, CUDNN_CONVOLUTION, convdims,
 
 const CUDNNFloat = Union{Float16,Float32,Float64}
 
-function cudnnConvolutionDescriptor(cdims::DenseConvDims, x::DenseCuArray{T}) where T
+function cudnnConvolutionDescriptorAndPaddedInput(cdims::DenseConvDims, x::DenseCuArray{T}) where T
+    # The main purpose of this function is to catch asymmetric padding which cudnn does not support
+    # If we find asymmetric padding we'll make a copy of x which is manually padded so that we can
+    # call cudnn with symmetric padding.
+    pad = collect(NNlib.padding(cdims)) # work with an array to make things more type stable
+    all(pad[1:2:end] .== pad[2:2:end]) && return (cudnnConvolutionDescriptor(cdims, x), x, identity)
+
+    # Maybe we should warn the user that this copies data, but other ML libs generally don't warn
+    sdims = NNlib.spatial_dims(cdims)
+    
+    # Naive implementation, is there a faster way?
+    # How much we need to pad x manually: The absolute difference between pad_left and pad_right, pad_top
+    # and pad_bottom etc. respectively. We keep the sign here though because we use it below to figure out
+    # which side of x to pad.
+    pad_manual = pad[1:2:2sdims] .- pad[2:2:2sdims]
+    # How much we can let cudnn pad: The smallest padding amount between pad_left and pad_right, pad_top 
+    # and pad_bottom etc. respectively
+    pad_cudnn = min.(pad[1:2:2sdims], pad[2:2:2sdims])
+
+
+    x_padded = similar(x, (size(x)[1:sdims] .+ abs.(pad_manual))..., size(x)[end-1:end]...)
+    # We could do the same yucky indexing stuff for the zeros too so we don't have to write zeros in the whole array.
+    # Not sure if it is worth it though...
+    fill!(x_padded, 0)
+    # This is a bit yucky, but we are basically figuring out where in x_padded we shall insert x_inds
+    # Haven't benchmarked if this has any advantages over a more readable solution, e.g. writing dim by dim in a loop 
+    x_inds = range.(1 .+ max.(0, pad_manual), size(x)[1:sdims] .- min.(0, .-pad_manual))
+    x_padded[x_inds..., :, :] = x
```

Comment:
Would cartesian indices and/or a view + loop be less yucky here?

Reply:
(I don't have a concrete idea of what either would look like, just speculating)

Reply:
I always keep forgetting that `CartesianIndex` exists, thanks for reminding. Not sure if they have other tricks than the autobroadcasted `max/min` but that came in handy here, although I did not feel it justified removing the comment about yuckyness :)

One could also trivially use setdiff to only set the padded elements to zeros, but that took the benchmarking time for a `32x32x128x64` sized input from 6.5 ms to 8 seconds :)